### PR TITLE
Modular Layout

### PR DESCRIPTION
**Pull request by HGuillemet**
Modified master pom and opencv sub-module to use the new package-targetting feature and allow modularisation.

### CODE REVIEW

**Comment by saudet** [comment]
This is looking good! Let's merge it after we get `module-info` "classes" in place so that people can start testing this with `jlink` on the snapshots?

**Comment by HGuillemet** [comment]
Adding of module-info now works for both java and native jars, using moditech. Snapshot version of moditech is needed (see https://github.com/moditect/moditect/issues/75), until beta2 is released.
The extra execution of the compiler with a jdk > 8 is not needed anymore.

**Comment by HGuillemet** [comment]
sur travis ? J'ai l'impression que c'est le snapshot moditect qui est compilé avec java 8+ et sur la plateforme qui passe pas tu utilises le jdk 7

**Comment by saudet** [comment]
So, we'll need to update that! Can it still compile code for Java 7 though?

**Comment by HGuillemet** [comment]
moditect doesn't compile with java 7. Is Java 8 available on ppc64le ? I guess it's not a problem to request java 8 pour the building of presets.

**Comment by saudet** [comment]
I'm sure it's available, but what I want to do know is, does it produce class files that are incompatible with Java 7? If we can't create libraries for Java 7 with ModiTect, we won't be able to use it.

**Comment by HGuillemet** [comment]
moditect only creates the module-info.class, so what problem are you thinking about ?
What is blocking Travis is that maven is run with JDK 7 on ppc64le and that the moditect maven plugin uses lambda expressions and needs Java 8 to run. So maven must run with JDK 8. This is a problem related to the building tools only.

**Comment by saudet** [comment]
Ok, sounds good then. Could you try to update to JDK 8 for linux-ppc64le as well?

File: opencv/pom.xml [hunk 1]
```diff
@@ -70,41 +71,47 @@
...
-                <exclude>org/bytedeco/javacpp/${javacpp.platform}${javacpp.platform.extension}/share/**/*.dylib</exclude>
-                <exclude>org/bytedeco/javacpp/${javacpp.platform}${javacpp.platform.extension}/share/**/*.so*</exclude>
-                <exclude>org/bytedeco/javacpp/${javacpp.platform}${javacpp.platform.extension}/x86/**/*.lib</exclude>
-                <exclude>org/bytedeco/javacpp/${javacpp.platform}${javacpp.platform.extension}/x86/**/*.dll</exclude>
-                <exclude>org/bytedeco/javacpp/${javacpp.platform}${javacpp.platform.extension}/x86/**/*.exe</exclude>
-                <exclude>org/bytedeco/javacpp/${javacpp.platform}${javacpp.platform.extension}/x64/**/*.lib</exclude>
-                <exclude>org/bytedeco/javacpp/${javacpp.platform}${javacpp.platform.extension}/x64/**/*.dll</exclude>
-                <exclude>org/bytedeco/javacpp/${javacpp.platform}${javacpp.platform.extension}/x64/**/*.exe</exclude>
-                <exclude>org/bytedeco/javacpp/${javacpp.platform}${javacpp.platform.extension}/**/*cascade*</exclude>
+                <exclude>lib/*.so*</exclude>
```
Comment (saudet) [inline opencv/pom.xml]: I don't think we need to change any of those directories anymore. Could you try to revert all those changes back?

Comment (HGuillemet) [reply opencv/pom.xml]: Right. Done.

Comment (saudet) [reply opencv/pom.xml]: It's still the same. Let me know when you've reverted the paths here.

File: pom.xml [hunk 2]
```diff
@@ -164,7 +165,7 @@
             </goals>
             <configuration>
               <includes>
-                <include>org/bytedeco/javacpp/presets/*.java</include>
+                <include>org/bytedeco/javacpp/**/*_presets.java</include>
```
Comment (saudet) [inline pom.xml]: I think we can simplify this by just not compiling what's in `${project.basedir}/src/gen/java`. Can you try that?

File: opencv/pom.xml (hunk 1 above)
Comment (saudet) [reply opencv/pom.xml]: Can you either revert the changes to these excludes, or remove them if you find that they are no longer necessary?

File: opencv/src/main/java9/module-info.java [hunk 3]
```diff
@@ -0,0 +1,41 @@
+module org.bytedeco.javacpp.opencv {
```
Comment (saudet) [inline opencv/src/main/java9/module-info.java]: We'll also need to add `org.opencv` to the list.

File: opencv/src/main/java/org/bytedeco/javacpp/cvkernels.java [hunk 4]
```diff
@@ -1,98 +0,0 @@
...
-import static org.bytedeco.javacpp.Loader.*;
-import static org.bytedeco.javacpp.opencv_core.*;
-
-/**
- *
- * @author Samuel Audet
- */
-@Properties(inherit=opencv_core.class, value={
-    @Platform(define={"MAX_SIZE 16", "CV_INLINE static inline"}, include="cvkernels.h", compiler="fastfpu") })
-public class cvkernels {
```
Comment (saudet) [inline opencv/src/main/java/org/bytedeco/javacpp/cvkernels.java]: We'll need to add back this file as well. It's a few custom operations that I've added before that are used in JavaCV...

File: pom.xml [hunk 5]
```diff
@@ -342,21 +338,66 @@
...
+              <archive>
+                <manifestEntries>
+                  <MultiReleaseOutput>true</MultiReleaseOutput>
+                </manifestEntries>
+              </archive>
               <excludes>
-                <exclude>org/bytedeco/javacpp/windows-*/*.exp</exclude>
-                <exclude>org/bytedeco/javacpp/windows-*/*.lib</exclude>
-                <exclude>org/bytedeco/javacpp/windows-*/*.obj</exclude>
+                <exclude>/windows-*/*.exp</exclude>
```
Comment (saudet) [inline pom.xml]: Same here, are these excludes still required? I think they are, so let's at least revert the changes so they are effective.

File: pom.xml [hunk 6]
```diff
@@ -285,8 +286,9 @@
               <goal>build</goal>
             </goals>
             <configuration>
+              <outputDirectory>${project.build.directory}/native/org/bytedeco/javacpp/${javacpp.platform}${javacpp.platform.extension}</outputDirectory>
```
Comment (saudet) [inline pom.xml]: This isn't going to work for Android though. We need to place the libraries in `${javacpp.platform.library.path}` for that platform. We should be able to use the profiles below in the pom.xml file for that though. Could you give this a try?

File: opencv/pom.xml (hunk 1 above)
Comment (HGuillemet) [reply opencv/pom.xml]: Reverted

File: pom.xml (hunk 2 above)
Comment (HGuillemet) [reply pom.xml]: This prevent for compiling helper classes too (that may depend on generated classes that will be generated later in the build process)

File: opencv/src/main/java9/module-info.java (hunk 3 above)
Comment (HGuillemet) [reply opencv/src/main/java9/module-info.java]: Right. Added.

File: pom.xml (hunk 5 above)
Comment (HGuillemet) [reply pom.xml]: Ok. Reverted.

File: pom.xml (hunk 2 above)
Comment (saudet) [reply pom.xml]: The helper classes are not part of the generated classes. It's not very important, but if the generated classes contain syntax errors, this prevents the build from passing until we erase them manually, so we shouldn't try to compile them at this point.

File: pom.xml (hunk 6 above)
Comment (saudet) [reply pom.xml]: Should be enough, let's see what that gives, thanks!

`javacpp.platform.library.path` isn't available for any other platform. If neither it nor `outputDirectory` is set, the effective output directory gets based on the scanned classes, so we can't really set something like `javacpp.platform.library.path` before actually parsing everything in those cases.

File: pom.xml (hunk 2 above)
Comment (HGuillemet) [reply pom.xml]: We don't. At this point `compileSourceRoot` is not set, so defaults to `src/main/java` and the `includes` limits the source file to the presets file. Thus neither the generated classes nor the helper classes are compiled. I don't understand what's wrong in this configuration.

Comment (saudet) [reply pom.xml]: Ah, yes, it's not part of the same execution. I got mixed up. It looks alright.

File: pom.xml [hunk 7]
```diff
@@ -342,12 +339,13 @@
             </goals>
             <configuration>
               <classifier>${javacpp.platform}${javacpp.platform.extension}</classifier>
-              <skipIfEmpty>true</skipIfEmpty>
-              <includes>
-                <include>${javacpp.platform.library.path}/</include>
-                <include>${javacpp.platform}${javacpp.platform.extension}/</include>
-                <include>org/bytedeco/javacpp/${javacpp.platform}${javacpp.platform.extension}/</include>
-              </includes>
+              <classesDirectory>${project.build.directory}/native</classesDirectory>
```
[the remark this answers is not in the document]
Comment (HGuillemet) [reply pom.xml]: `native/` will only contain the output of the javacpp compiler. So no more need to filter out .java and various outputs from other building stage and that's why I removed it. But if you think normal situations may arise where, after some crosscompiling and no cleaning up, there are compiler outputs for different platforms in the directory, we might bring this includes back.

[the remark this answers is not in the document]
Comment (HGuillemet) [reply pom.xml]: Ok, added `<include>${javacpp.platform.nativeOutputPath}/</include>`, that should be enough.

File: pom.xml [hunk 8]
```diff
@@ -392,6 +437,7 @@
             <configuration>
               <minmemory>128m</minmemory>
               <maxmemory>1024m</maxmemory>
+              <sourcepath>${project.basedir}/src/gen/java:${project.build.sourceDirectory}</sourcepath>
```
[the remark this answers is not in the document]
Comment (HGuillemet) [reply pom.xml]: I don't understand how maven retrieves the `compileSourceRoots` from a specific execution configuration of the compiler, but it seems to do. I get the warnings too when I remove `sourcepath`. I suggest to add the path to org.opencv source path by overriding `sourcepath` in `opencv/pom.xml`, just like `compileSourceRoots` is overridden to add this path.

**Comment by saudet** [comment]
Let's not add any extra settings if we don't need to. Does it work better
if we specify the directories manually? If not, let's remove it!

**Comment by HGuillemet** [comment]
Yes, no more warnings.

**Comment by HGuillemet** [comment]
> Have you also tried changing the compileSourceRoots in the same order?

Yes, no changes.
